### ISSUE TITLE
Seeing 100% CPU on random times, stability issues

### ISSUE THREAD

**Issue by kitkit201** [haproxy/haproxy#1881]
### Detailed Description of the Problem

Hello Haproxy users,

We have been seeing our 20 or so HAproxy nodes spike up to 99% CPU on the cluster at random times which is causing our site to stop taking traffic for about 10 minutes of time and causing our Autoscaling group in AWS to spike up to our maximum of 40 nodes to cover the traffic.

We are using a recent version of HAproxy (2.4.17) . We are not restarting the systems or processes at any time other than when it starts up by itself in our config. We use AWS to Autoscale hosts at 20% cpu (which seems low, but we know that we overstress HAProxy too much, it seems to topple over) . Perhaps we are doing something wrong with our configs and it’s not optimal or something?

We use a minimum of 20 c5.xlarges and a maximum of 40 c5.xlarges with a CPU tracking of 14% Autoscaling when the Autoscaling group hits that threshhold

### Expected Behavior

We are expecting the CPU behavior to be higher because we are seeing anything higher than CPU 14% for the hosts to topple and get out of control. That seems low in terms of CPU 

### Steps to Reproduce the Behavior

Run the config as stated

### Do you have any idea what may have caused this?

This was our default config while we were building things 

### Do you have an idea how to solve the issue?

Possibly connections or some system level settings

### What is your configuration?

```haproxy
global
log 127.0.0.1 local6
maxconn 30720
user haproxy
group haproxy
daemon
quiet
pidfile /var/run/haproxy.pid
tune.maxrewrite 16384
tune.bufsize 65536
nbproc 1
nbthread 8
stats socket /var/run/monitor-haproxy/haproxy.sock mode 0600 level admin

defaults
log global
mode http
option httpchk GET /health/
option httplog
option dontlognull
option dontlog-normal
option log-separate-errors
retries 3
option redispatch
maxconn 30720
timeout connect 5s
timeout client 60s
timeout server 60s
option http-server-close
option tcp-smart-accept
option tcp-smart-connect
option allbackups
stats enable
stats show-legends
stats uri /lbstats
stats refresh 10
stats realm Florin
stats auth username:passsssword1234
balance roundrobin
hash-type consistent djb2 avalanche
no option log-health-checks

resolvers localdns
nameserver localdns 127.0.0.1:53

frontend stats
bind *:7684
stats enable
stats refresh 10s
stats realm Monitor
stats uri /hc

frontend http-in
bind *:80
default_backend xxxxxxx

frontend https-in
bind *:443 ssl crt /etc/ssl/private/foo.bar`

A bunch of acl to identify traffic
4 backends to distribute traffic based off of acl readings
backend xxxxx
option httpchk GET /favicon.ico HTTP/1.1\r\nHost:\ 10.78.0.0
server-template xxxxxxx 1-12 internal-xxxxxx.us-east-1.elb.amazonaws.com:443 resolvers localdns resolve-prefer ipv4 init-addr last,libc,none check ssl verify none no-sslv3 inter 10s fall 4

backend xxxxx-xxxxxxx
timeout queue 10s
option httpchk GET /favicon.ico HTTP/1.1\r\nHost:\ 10.78.0.0
server-template ABC 1-12 internal-xxx.us-east-1.elb.amazonaws.com:443 resolvers localdns resolve-prefer ipv4 init-addr last,libc,none check ssl verify none no-sslv3 inter 10s fall 4 maxconn 200
http-response return status 429 default-errorfiles if { status 503 }

backend xxxxxxx
option httpchk GET /status HTTP/1.1\r\nHost:\ 10.78.0.0
server-template ABCD 1-12 internal-xxxx.us-east-1.elb.amazonaws.com:443 resolvers localdns resolve-prefer ipv4 init-addr last,libc,none check ssl verify none no-sslv3 inter 10s fall 4

backend xxxxxxx-bots
timeout queue 10s
option httpchk GET /status HTTP/1.1\r\nHost:\ 10.78.0.0
server-template ABCDEFC 1-12 internal-xxxxx.us-east-1.elb.amazonaws.com:443 resolvers localdns resolve-prefer ipv4 init-addr last,libc,none check ssl verify none no-sslv3 inter 10s fall 4 maxconn 200
http-response return status 429 default-errorfiles if { status 503 }
```

### Output of `haproxy -vv`

```plain
$ haproxy -vv
HAProxy version 2.4.17-9f97155 2022/05/13 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.17.html
Running on: Linux 5.4.0-1061-aws #64~18.04.1-Ubuntu SMP Fri Dec 3 17:59:13 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -m64 -march=native -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ZLIB=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.5.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

### Additional Information

We are currently on Haproxy 2.4.17 . Happy to run to the latest version, but we had the same problem in 2.4.7 and upgraded to the 2.4.17 recently. I posted it in https://discourse.haproxy.org/t/seeing-100-cpu-on-random-times-stability-issues/7980 and it was recommended to post on here

**Comment by wtarreau** [comment]
I'm not seeing anything wrong in your config, which is pretty standard, and clean overall. What you're observing is at least not a deadlock nor a spinning loop, otherwise the watchdog would trigger after 1 to 2 seconds max to kill the process with a panic message. It's not an uncaught event either (e.g. shutdown or error), otherwise the process would continue to handle traffic while eating CPU. Thus what I suspect is that the problem revolves around the accept() code. Maybe the listener is disabled at some point then enabled again. Or maybe accept() fails, which would explain why the process cannot accept traffic anymore. Another possibility would be that the nf_conntrack module is loaded on the system and running with the default settings (very low) and that at some point it's not possible to establish outgoing connections anymore. The effect would be a waste of CPU with many connections being retried in parallel, and a difficulty in establishing new connections. In this case, issuing `dmesg` on the system would report "conntrack table is full". If/when the problem happens again, running `strace -f -c -p $(pidof haproxy)` for a few seconds then pressing Ctrl-C would report the attempted syscalls. If you feel brave, taking a small snapshot of these syscalls would be nice: `strace -ftts200 -o trace.log -p $(pidof haproxy)`. Be careful, do not share that last one publicly, as it may reveal some sensitive info (domain names, IP addresses, etc). It would also be interesting to know if you can still connect to your stats socket when this happens. If it works, then issuing `show info; show activity; show fd; show sess all` there would be extremely useful. Similarly, some info will be disclosed by the `show sess all` one, so you may want to issue it in a separate command and send it into a separate log file.

**Comment by kitkit201** [comment]
Thanks for the reply @wtarreau ! Does increasing the `nbthreads` help? We currently are using an AWS C5.xlarge which is 4 VCPU, but thre are only 2 physical cores. We had `nbthreads=8`  before and changed it to `nbthreads=4` just yesterday.

I don't know what is "normal" for a haproxy node, so i'm kinda driving blind, but judging by your comments above, the CPU metric seems to be within spec. Is that correct?

Thanks

**Comment by wtarreau** [comment]
You must never set nbthreads to a value higher than the number of CPUs visible to the system (4 in your case, as reported by nbproc). Nevertheless, haproxy normally issues a warning nowadays when it detects this. Anything above will cause extreme contention (but will not prevent it from accepting connections, it's just that it will be much slower and will eat a lot of CPU). And if you don't set anything it will pick the correct value. With 4 threads you already have plenty. In general it's not needed to set large values. The vast majority of use cases fit very well on a single one. To give you an idea I'm currently running tests at 2000 requests/s on a tiny sub-1watt computer powered over a USB plug. When you're dealing with SSL, compression, Lua, many rules, stick-tables, cache etc... all these add up to the processing and it can make sense to add threads to give extra processing power. But generally you'll quickly figure what's the limit for a thread and how it degrades when the limit is reached. It's good to never reach saturation though, in order to maintain low latencies. You can have a look at "show activity" on the stats socket, there's a running average "avg_loop_us" which indicates the measured internal latency in microseconds. It roughly indicates the processing delay that haproxy adds on a request-response round trip. It's desirable that it remains under one millisecond (1000 microseconds), ideally 100 microseconds for high traffic sites. If it's too high, it's when adding a few threads can help.

**Comment by kitkit201** [comment]
We just had another 10 mins of outage just now and i went onto some of the hosts to pull up data. CPU went to 99% on the existing hosts and ASG autoscaling had to kick in to add more HAProxy nodes. The `avg_loop_ms` seems close to 100 

`$ echo "show activity" | sudo socat unix-connect:/var/run/monitor-haproxy/haproxy.sock -
thread_id: 4 (1..4)
date_now: 1664357536.288355
ctxsw: 1274109777 [ 318710324 319203269 318315564 317880620 ]
tasksw: 222090922 [ 55572638 55632150 55476033 55410101 ]
empty_rq: 3170322 [ 1023609 738956 905946 501811 ]
long_rq: 1 [ 0 1 0 0 ]
loops: 507383075 [ 127118547 127057006 126862437 126345085 ]
wake_tasks: 153080586 [ 38321339 38341263 38235030 38182954 ]
wake_signal: 0 [ 0 0 0 0 ]
poll_io: 367035140 [ 91977090 91915389 91768151 91374510 ]
poll_exp: 2329 [ 91 1414 422 402 ]
poll_drop_fd: 113823 [ 28420 28350 28597 28456 ]
poll_dead_fd: 0 [ 0 0 0 0 ]
poll_skip_fd: 0 [ 0 0 0 0 ]
conn_dead: 0 [ 0 0 0 0 ]
stream_calls: 170555555 [ 42674071 42725685 42600616 42555183 ]
pool_fail: 0 [ 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 ]
cpust_ms_tot: 17100 [ 3696 4288 4297 4819 ]
cpust_ms_1s: 0 [ 0 0 0 0 ]
cpust_ms_15s: 0 [ 0 0 0 0 ]
avg_loop_us: 78 [ 85 75 76 75 ]
accepted: 1843122 [ 1061338 178630 501530 101624 ]
accq_pushed: 1843122 [ 460908 460136 461416 460662 ]
accq_full: 0 [ 0 0 0 0 ]
accq_ring: 0 [ 0 0 0 0 ]
fd_takeover: 0 [ 0 0 0 0 ]`

Here is a graph of what is happening. Nothing is jumping out at me other than thinking of lowering the CPU autoscaling target so we have more cpu capacity in reserve. What are your thoughts?

<img width="1328" alt="Screen Shot 2022-09-28 at 2 38 13 AM" src="https://user-images.githubusercontent.com/5623442/192746239-9a0e9339-a812-45ae-a02a-b09c84a8e5ee.png">

**Comment by wtarreau** [comment]
Sorry @kitkit201  I hadn't notice you had responded! Your "show activity" output looks normal, there's not even any imbalance between threads. This makes me think that we should also emit the uptime there by the way, that would help spot abnormally high numbers. Even the ratio of poll_io to accepted connections looks reasonable (~200 I/O per connections so nothing excessive if you're having sufficient keep-alive connections).

I'm wondering if in fact you could be facing a situation where the frontend's maxconn is reached (or the process' one) and the listener would spend its time enabling and disabling polling. This could be expensive and could also explain why all threads are seeing the same stats. After all, the frontend sessions seem to have shown a peak at this moment, it could match. The output of "show info" during or after the event would reveal this.

**Comment by yard** [comment]
Hey @wtarreau! We are running haproxy 2.7.1 (using official docker image) and observe very similar symptoms: a cluster of 3 haproxy works just fine handling the varying amounts of traffic, discovering new / discarding retired servers and consumes a very modest amount of CPU. At random times, however, the CPU usage spikes, so does the amount of backend re-dispatches, average queue time and other metrics – without any visible change in incoming traffic, backend performance and so on. We have observed that behaviour many times, tried different versions etc but without any luck.

Our configuration file looks like:
```
global
  log stdout local0 notice
  log-send-hostname
  maxconn 65535
  user haproxy
  group haproxy
  daemon
  nbthread 4
  cpu-map 1/all 0-3
  ssl-default-bind-options no-sslv3
  stats socket /tmp/haproxy.sock mode 0600 level admin
defaults
  balance roundrobin
  log global
  mode http
  option redispatch
  option dontlognull
  option forwardfor
  option http-keep-alive
  timeout connect 5000
  timeout client 50000
  timeout server 50000
listen stats1
  bind :19361
  mode http
  stats enable
  timeout connect 10s
  timeout client 1m
  timeout server 1m
  stats hide-version
  stats realm Haproxy\ Statistics
  stats uri /
  stats auth stats:stats
frontend port_443
  bind :443 ssl crt /usr/local/etc/certs/
  maxconn 65535
  use_backend SERVICE_collector-beta_asg-collector-beta
resolvers docker
  nameserver dns0 127.0.0.11:53
  resolve_retries 30
  timeout retry 2s
  hold valid 10s
  accepted_payload_size 8192
backend SERVICE_collector-beta_asg-collector-beta
  option http-keep-alive
  http-reuse always
  timeout queue 1s
  server-template collector-service 1-64 asg-collector-beta:80 resolvers docker resolve-prefer ipv4 check inter 2000 rise 2 fall 3 maxconn 8 init-addr none
```

The output of `show activity` during one of the incidents is as follows:
```
thread_id: 3 (1..4)
date_now: 1672147611.817002
uptime_now: 1033.417443
ctxsw: 216050614 [ 54157248 53822139 54061939 54009288 ]
tasksw: 75834208 [ 18994593 18957694 18953870 18928051 ]
empty_rq: 140651 [ 39866 41742 33093 25950 ]
long_rq: 198304 [ 48241 56696 47306 46061 ]
loops: 19989283 [ 5030140 4948321 5009587 5001235 ]
wake_tasks: 780982 [ 194064 201938 192319 192661 ]
wake_signal: 0 [ 0 0 0 0 ]
poll_io: 19937036 [ 5016821 4935506 4996455 4988254 ]
poll_exp: 2376 [ 578 591 628 579 ]
poll_drop_fd: 3808 [ 914 1045 961 888 ]
poll_skip_fd: 16587 [ 4229 4142 4095 4121 ]
conn_dead: 0 [ 0 0 0 0 ]
stream_calls: 62152774 [ 15569322 15532598 15537636 15513218 ]
pool_fail: 0 [ 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 ]
cpust_ms_tot: 3276 [ 786 783 850 857 ]
cpust_ms_1s: 0 [ 0 0 0 0 ]
cpust_ms_15s: 5 [ 0 0 0 5 ]
avg_loop_us: 597 [ 636 606 578 567 ]
accepted: 379232 [ 96053 113463 88128 81588 ]
accq_pushed: 379232 [ 94984 94747 94528 94973 ]
accq_full: 0 [ 0 0 0 0 ]
accq_ring: 0 [ 0 0 0 0 ]
fd_takeover: 21730 [ 5569 5442 5421 5298 ]
```

P.S. Happy to open a separate issue if that's more appropriate.

**Comment by wtarreau** [comment]
@yard thanks for your report. For now it's hard to know if it's the same issue so better keep them together. your ctxsw rate seems high for the uptime (~200k sw/s for ~380 conn/s). The calls to process_stream() look high as well (62k/s). What's you're rough request rate ? Is it in the hundreds, thousands or tens of thousands of requests per second ? Also, the average loop time is about 600 microseconds, or more than half a millisecond. I'm wondering if you're not just witnessing a regular attack on TLS. Could you please run "perf top" when the issue happens and paste the output here ? It will indicate where the CPU is spending its time. We could see a lock, some SSL code, regex code etc which usually help.

**Comment by yard** [comment]
Hey! Yeah so the request rate is around 10k per second in this example, and I am pretty sure it’s not a TLS attack (we are behind Cloudflare, so chances are extremely low). It’s worth noting that this very same setup does handle 2x as much without hiccups for days (I did try that by removing other instances to get more traffic flowing through this one).

I will try to catch it again and run perf top ofc to provide more input.

P.S. Thank you for such a swift turnaround and kudos on an outstanding tech you are building!

**Comment by wtarreau** [comment]
OK indeed all of this makes sense. I asked because sudden spikes like this can sometimes correspond to an attack. 6 process_stream() calls per request are reasonable. Did you eliminate some elements in your config that could occasionally represent heavy processing (e.g. some `map_regm` converters involving many regex etc) ? This could explain the high loop time (for me 600 microseconds is on the high range). Next time it happens, please also issue a "show info". I'm interested in seeing the number of running tasks among other things.

**Comment by yard** [comment]
Re `map_regm` and stuff - the config I posted is not redacted, that is literally all we have, so no regex or any logic whatsoever, basically straight dispatch to a backend after SSL termination. Only “non-convetional” (very arguably ofc) bit is aggressive http reuse to keep connections to backends persistent since all the requests we handle are one-off isolated ones without any stickness requirements etc.

I will run both `show info` and `perf top` as soon as I manage to catch the issue again (since it is pretty sporadic, might take days until it appears again). Anything else to take a note of during the incident? Maybe connection rate to backend servers or smth?

**Comment by wtarreau** [comment]
I'm wondering if it could correspond to a missed error or shutdown. Maybe it lasts for the duration of a timeout only. Among the things you can try to catch, I'm thinking about a few things to collect during the event and also once it's not there (in order to compare):
- show actiity
- show tasks
- show info
- perf top

And when the problem happens, it will be useful to also have:
- show fd
- show sess all

Warning, the last one *will* disclose some private information. I suggest to issue `set anon on` before `show sess all` so that IP addresses and object names are hashed in the output. The last two commands above will provide large dumps, so make sure to redirect them directly into a file. Please have a quick look at their contents before shaering, to make sure nothing sensitive is leaked. With a bit of luck we'll find a stuck connection (front or back) that is shut but doesn't want to go and that could explain the problem.

Thanks!

**Comment by kitkit201** [comment]
Hi @wtarreau , Happy New Year. Hope you are doing well .

We are still seeing the issue. We upgraded to HAProxy 2.7.1 before the holiday break and it was going fine until about a few days back (Jan 7th, 2023) where we saw the same issue of happen. 

Long story short, I wasn't able to grab the info you requested, but I will write a script so i can run it when the time happens.

**Comment by wtarreau** [comment]
There's something very useful you could also do when the problem happens, which is to enable task profiling on the CLI: `set profiling tasks on` then wait a few seconds and issue `show profiling tasks`. It will report tasks that are called a lot, this should give us some hints. And if you're willing to rebuild haproxy, please pass `DEFINE=-DDEBUG_TASK` to your "make" command line. It will provide very useful caller information in "show profiling" that may help figure the cause of this activity.

**Comment by yard** [comment]
Hi @wtarreau! Happy NY to you and your team!

I was away for 2 weeks, but my colleague managed to capture the output you requested – attaching this here (we skipped  `set anon on` command since IP addresses aren't of our clients anyway).

[haproxy-2.7.1-logs.zip](https://github.com/haproxy/haproxy/files/10390981/haproxy-2.7.1-logs.zip)

**Comment by wtarreau** [comment]
Hi,

thanks for this. Do you have a copy of the perf top that is not truncated ? This one is zero byte. Also, are these the captures during the event or without the event ? If you see it again, really do not hesitate to try the task profiling above, it could speed up the location of the problem.

**Comment by wtarreau** [comment]
What I can see at least is that no stream is looping. Among the 286 tasks, ~83% of them are in I/O callbacks. This could be normal but still sounds suspicious, especially for the ssl one, given the reasonably modest ssl conn rate. So it could be possible that a failure is not properly caught here. We'll have to scan all the connection flags to try to see.

**Comment by yard** [comment]
hey, sorry, re-attaching `perf top` output (captured with `perf record`, so can probably truncate more/less, just lmk).
[perf-record.txt](https://github.com/haproxy/haproxy/files/10391709/perf-record.txt)

the captures were during the event (high cpu, many dropped requests – just like before), this one lasted for couple hours in fact, but generally it resolves in circa 10 minutes (with no manual intervention or severe traffic changes).

**Comment by kitkit201** [comment]
Good morning. So we just got another blip about 30 minutes ago. I've attached the logs and it's rather similar to what @yard was seeing with the "show profiling tasks" with the I/O Callbacks.

[haproxy-2.7.1-logs-kitkit201.txt.zip](https://github.com/haproxy/haproxy/files/10392294/haproxy-2.7.1-logs-kitkit201.txt.zip)


Of note, we are only seeing these high aborted transfers from our "Hermes" nodejs backends (and not the other 3). Its hard for me to tell if this is caused by high number of requests leading to retries or real requests (likely the former) Here's a screenshot of some haproxy metrics at the time of the incident. 
<img width="1824" alt="Screenshot 2023-01-11 at 4 31 43 AM" src="https://user-images.githubusercontent.com/5623442/211808874-801531fe-fd1a-4d37-b969-b0689750b02d.png">

**Comment by wtarreau** [comment]
There's something odd on this screenshot. It shows that the servers went down for several minutes, that request time has exploded (so typically the server(s) totally disappeared from the network and requests had to wait for the timeout), the session count has increased but to small values, not even to the point of triggering the queue. But my understanding is that the current sessions (157.7) corresponds to the position of the current cursor on the graph, which is so low that it's merged with the border, which would indicate that the peak it very high, in the thousands. Yet there is no queue! Are these servers also configured with "maxconn 200" like the rest of the config ? Or am I misreading the graphs (with no scale it's really not easy) ? I'm seeing a peak of 7k req/s during the incident, which makes me think that in case of errors, clients simply retry immediately.

Then the "CPU issue" could very well be just the following sequence:
  - clients sending requests to a dead server
  - haproxy returning 503 because the server is dead
  - client retrying immediately instead of waiting a bit

This would result in an increase of the request rate which could justify a CPU usage increase. Of course the request rate is not big, but it obvsiouly depends on the processing that's being done per-request.

The captures in the file all indicate that everything is doing pretty well. There are no errors, the process is even at 97% idle so not doing anything, and the traffic looks good. That's very likely be because the incident is gone and everything has recovered, but it could confirm that we need to search for such a behavior instead.

I'm thinking about something though. You're having a 30k limit on the concurrent connections. During such an incident it's very likely that clients accumulate and that the limit is reached. When this is the case, the limitation will be quickly applied and released on a per-connection basis, and this can suddenly become very expensive for the CPUs since they wake up with traffic, have to be paused, then resumed once a connection is released, etc. It would be interesting to look at the same period for other busy frontends or backends and see if you instead observe a drop of the request rate or concurrent requests. This could then confirm such a hypothesis.
On the stats page there's column "sessions/max" which reports per-frontend the highest number of concurrent connections that was seen (we don't have it per process, as I thought). If you see very high numbers, that could also fuel that hypothesis. One test to do then is to significantly increase the global `maxconn` value (e.g. triple it, and be aware that this will increase memory usage), and see if the problem persists.

**Comment by kitkit201** [comment]
@wtarreau thanks for the quick response back. Yes, i took the logs 10 minutes after the event so it wasn't during the event. i'll do better next time at capturing it as it happens :) 

So what's happening on my side is when the event happens, my AWS Autoscaling increases because the CPU on the cluster is over the average of 12% and it keeps adding new HAProxy nodes until its back within range. It went from about 26 HAProxy nodes to about 61 nodes during this event.  (Perhaps CPU isn't a great gauge for autoscaling in this case either). 

In terms of the hermes backend service, it did increase CPU and tasks count from 22 to 34 to keep up with the CPU autoscaling. 

So really the question is tier is causing which pain. I do like the idea of increasing the concurrent connections count on haproxy so i will be changing it from 30k to 120k once i get a chance to later today.

**Comment by yard** [comment]
> hey, sorry, re-attaching `perf top` output (captured with `perf record`, so can probably truncate more/less, just lmk). [perf-record.txt](https://github.com/haproxy/haproxy/files/10391709/perf-record.txt)
> 
> the captures were during the event (high cpu, many dropped requests – just like before), this one lasted for couple hours in fact, but generally it resolves in circa 10 minutes (with no manual intervention or severe traffic changes).

Hi @wtarreau! Anything that catches your eye in the attached `perf record` dump?

**Comment by wtarreau** [comment]
Sorry, I thought I had responded to this one but can't see it, probably lost in another tab :-(
What I noticed in this dump is that your CPU is mostly spent in looking up iptables/nftables connections. That amount of CPU is not negligible, it makes me think that you could be facing a large connection count with too small a hash table. You should check `/sys/module/nf_conntrack/parameters/hashsize` to make sure that it's large enough (I generally encourage in having it set to 1/4 of the total expected connections count, including time_waits, that you can probably find in `/proc/sys/net/netfilter/nf_conntrack_count`.

**Comment by yard** [comment]
Hey, so I have just checked `/sys/module/nf_conntrack/parameters/hashsize` and it's set to 65536 and `/proc/sys/net/netfilter/nf_conntrack_count` sits at ~20000 – but at the moment there's no incident happening, shall I check those once we see the glitch occurring? 

I don't completely understand the theory though: if the hash table is too small, why would it manifest the issue at random times (not correlated to highest traffic etc)? E.g. we have seen the issue manifesting itself at 8k RPS and then not having it for days while traffic reached as high as 15k RPS for extended periods of time...

**Comment by wtarreau** [comment]
The effect of too small a table could happen when there is a very large number of connections. Let's say you're receiving attacks from time to time for example and end up with 1M connections instead of 20k, the lookup time will be multiplied by 20. But with 64k entries there's little risk anyway. Another thing that can consume a lot of CPU is searching for a spare source port to connect to a server when all are in use, and this can also happen if outgoing connections are in TIME_WAIT on the haproxy side, something we try very hard to avoid (and normally they do never happen). Then the connect() syscall can take ages to try different ports until it finds one, and this is purely CPU usage. When this happens you instantly see it in perf top. Such a problem could be a side effect of a dyning server (e.g. connections are established and closed very fast).  With that said, if the server gets slow and you run at 8k rps, in the worst case it can take as little as 8 seconds to fill all source ports (and even less if you're limited to the default 32768-61000). And again trying to find a spare source port can become quite expensive. It could be useful to look at the "backend session max" in your stats to see what is the highest concurrency ever seen on the backend. This could explain the fact that the CPU is used at the same time the server goes down. Please have a look at your outgoing source ports range in `/proc/sys/net/ipv4/ip_local_port_range` and if it's the default, it might be worth extending it to 1024-65535.

**Comment by yard** [comment]
Ok so re `backend sessions max` – we have specifically limited that and it normally never goes over a few dozens connections per backend, **but** during the incident (during which the logs where collected) it did spike to thousands:
<img width="434" alt="Screenshot 2023-01-12 at 14 57 28" src="https://user-images.githubusercontent.com/62775/212100295-ea6c0331-0a2a-4e9c-ad22-e7d2a9b2224a.png">

I will try to extend local port range and will revert back if the issue returns.

**Comment by kitkit201** [comment]
Small update: We tried changing the max connections from 30k to 100k and still had those blips. Meanwhile, I wrote a script to trigger those logs when the situation happens again based off of cpuload.

**Comment by yard** [comment]
Hey @wtarreau! So we have ran the updated setup for a week now and did notice an occurrence of the event once during this period. The event was rather short (under 5 minutes) and did manifest itself as:

1. spike in queue time
<img width="434" alt="Screenshot 2023-01-20 at 10 35 33" src="https://user-images.githubusercontent.com/62775/213663062-a06b2703-c814-43a6-90a1-87cdeada7e8a.png">

2. increased amount of 503 codes returned by haproxy
<img width="434" alt="Screenshot 2023-01-20 at 10 36 47" src="https://user-images.githubusercontent.com/62775/213663325-1d3e583b-fe76-4e43-9ba2-7772f833e2d9.png">

3. increased number of backend sessions
<img width="434" alt="Screenshot 2023-01-20 at 10 37 15" src="https://user-images.githubusercontent.com/62775/213663445-96f5b10f-a44c-4a27-8683-51ab0e24691d.png">

Just like before,  The CPU spike was very minor (although existent), so I start to suspect the higher CPU load is really just a symptom rather than a cause of sorts.

Now we are in a funny realm where we are actually unsure how to debug this problem :) Would be glad to hear if you have any ideas on what to look into during/before/after the incident – will try to automate that and come back with a more concise report.

**Comment by wtarreau** [comment]
So indeed, the CPU usage you're seeing is a result of this, most likely the fact that when you are getting close to the maxconn value, some LB algos need to skip saturated servers to find another one, and this increases the time needed to find a usable server, hence inter-thread contention thus CPU usage.

At this point what is important is to dig in the logs, trying to see request by request how the response time increases, or if there is a moment during which there is no response at all. If you see the response time increase it can be caused by a parasitic activity on the server (e.g. backups, log compression, swapping). If you see a sudden stop of any response, it can be caused by something more violent such as a garbage collector, or a reboot of a stateful equipemnt on a set of persistent connections (database, ldap etc) that need to time out first. But in any case, the logs will be your friend here.

**Comment by yard** [comment]
Well it actually looks more "internal" to haproxy rather than our servers. The latest incident (and many before that) have actually occurred during _lower_ traffic conditions:
<img width="435" alt="Screenshot 2023-01-20 at 11 31 31" src="https://user-images.githubusercontent.com/62775/213674250-1c14634c-5531-46cb-982f-0abc1a37f0e1.png">

In addition to that, the traffic is homogenous, the set of backend servers is static etc etc – it does pop out randomly and is not connected to any other event. Obviously, nothing interesting in the logs either. The only effect we observe on backend servers is slightly higher CPU on some and slightly lower on others (the change is from 30% to 25/35% correspondingly and probably connected to haproxy redispatching the request a few times and some servers getting more lucky than other, but it's definitely not overloading one or anything, so I guess it's a metric we can ignore). There is still, of course, a possibility of network glitch or a server meltdown or anything else we missed, but it's definitely not something obvious or noticeable on our monitoring.

**Comment by wtarreau** [comment]
It would be useful to have a look at network retransmissions and rx packet drops around such events. I've faced comparable issues in the past with a faulty NIC that has a few defect bits in its buffers, randomly causing excessive retransmits. Also make sure the machine haproxy is running on never has any swap used at all. Network losses on rx can be caused by temporary memory shortages and usually translate in connections timing out due to the SYN-ACK not being received.

**Comment by yard** [comment]
Wouldn't that bump time to connect to a backend metric? We saw jumps on that when we did experience networking issues, but this time it's flat. Swap is off and the spare amount of RAM is above 50% too.

**Comment by wtarreau** [comment]
Yes you're right normally you would also see the connect time increase in such cases, and the queue would only be a side effect of it. I didn't know if you were monitoring connect times since you only mentioned queue time :-)

However if you're having lots of persistent connections and never reopen connections, then all of this could indicate a slowdown either on the server or on the network. What would be important to see on your graphs is if the problem affects:
- all servers at once, regardless of backends => this would indicate that the problem lies within the machine hosting haproxy and its local network, including kernel and haproxy themselves ;
- all the servers from a backend at once => this would indicate a problem related to these servers, possibly a source port shortage, a problem on a specific network path, or some limits related to the backend's config
- only one server (or the same server in different backends all on the same IP), this would rather correspond to the server itself and/or the path to it that is not shared by other servers not experiencing the problem.

Also since you're on virtual instances, it would be nice to see the "cpust" lines in "show activity". Maybe you're seeing your CPUs stolen because there are too many people on the same hardware, and that causes traffic to be jerky. I don't see why it would cause excess queues but it could at least explain some latency spikes (trying to address one issue at a time).

**Comment by yard** [comment]
So we had a few instances of the issue happening today. Unfortunately, we missed the opportunity to gather performance logs this time (but this probably fine; I doubt we would see anything in new in there).

Re you question: the issue affects _random_ haproxy servers each time. For example, today it occurred on 1st and 3rd servers, 2nd was fine; previous one only affected 2nd server and so on.
<img width="1468" alt="Screenshot 2023-01-23 at 10 28 29" src="https://user-images.githubusercontent.com/62775/214005809-6dd40e18-d552-46b7-a4e1-3c84adfc2b6c.png">

Just like in other instances, no traffic spikes were present, no new servers added or old ones deleted; all backend servers were fine during the whole duration of the incident (the backend response time metric did jump during peak queue time, but it does seem to be a direct consequence of the increased queue time; backend servers themselves did not seem to be affected).

Re CPU steal – it's virtually non-existent (under 0.01%) and does not change before, during or after the incident:
<img width="1471" alt="Screenshot 2023-01-23 at 10 27 46" src="https://user-images.githubusercontent.com/62775/214005660-f69a13e0-688a-4e68-bac6-6bca77c6dc35.png">

We will try to add more instrumentation to backend servers to try identifying any suspicious patterns, but the way the charts look still hints at some glitch within haproxy itself.

**Comment by kitkit201** [comment]
Hi @wtarreau , Its been a while, but I have struggling to get live logs for you but finally got some logs for you that may help. From my side, i'm trying to determine where the issue lies, it's either the HAProxy, our Frontend or our Backend.. all 3 are likely suspects, but i think its something to do with either our Frontend or our Backend not being able to hadnle the connections .

**Comment by kitkit201** [comment]
Here is some better logs at the height of this incident. this has way more logs 
[kitkit201-debug-main.txt](https://github.com/haproxy/haproxy/files/10736968/kitkit201-debug-main.txt)

**Comment by wtarreau** [comment]
Awesmoe! At first glance, 28.5k tasks, 30k FDs and a loop time of 6.7ms really make me think that all the time it spent in the connect() syscall trying hard to find a spare source port for an outgoing connection. Please try to increase /proc/sys/net/ipv4/ip_local_port_range (e.g. 1024 to 65535) and see if that improves the situation. Also checking with "ss -an" that you don't have TIME_WAIT connection from haproxy to the servers will be useful. There should never be and if there are, they may participate to a source port shortage (on the other side, between clients and haproxy, it's OK).

**Comment by wtarreau** [comment]
Hmmm most of them are frontend connections, so it could be something between the client(s) and haproxy being limited, but it's not necessarily the case, haproxy could be behaving really badly. Still checking.

**Comment by wtarreau** [comment]
Hi @kitkit201 

yesterday we spend a few hours auditing all the code related to what you're seeing, and we could finally figure what is happening. For some reason (attack, slow servers etc) at some points you're very likely exhausting the source port range and connect() starts to get slow, explaining these 6.7 ms to complete a loop. The TLS handshakes in progress are in the run queue, but only one per loop is picked so as not to inflict large latencies. This means that you can do at best 150 handshakes/s per thread due to the high latency. This causes pending handshakes to take a very long time (48s for 28555) so users have long abandonned, and possibly clicked again. In the end, as they're processed in order in this tasklet queue, all those that are finally handled end up being useless, which results in no session, stream, request etc being processed. The handshakes are still there, and very likely they are the ones continuing to suffer from the connect() latency. Those in SSL also fail since their handshake also takes forever. Thus once you end up in this situation, traffic attempts hold the situation in this state. Most likely that unplugging the cable for one minute and replugging it would fix everything (don't do that, it's just to illustrate). And we've also found that the `long_rq` metric is wrong (it took us 3h to figure that the only case where it was possible and that could only result from an inconsistency between a flag and a list was not caused by them but by the incorrect metric).

The solution we've found to this is to permit to increase the amount of handshakes picked at once in the queue, to up to 1+1 per 1024 tasks when there are regular tasks, and 1+1 per 128 tasks otherwise. This means that in your case with such a load, handshakes would be processed at a rate of 56 per loop and per core instead of 1, thus covering for latency impacts and allowing the situation to recover. We've run long extensive tests on a large machine to make sure the ratios didn't degrade the normal behavior, and now that's OK, I'm still getting sub-millisecond response times at saturation.

Note that this will certainly not address the problem that it causing the high latencies in the first place, which is still to be determined, but at least it will make sure that such latencies don't cause a cascade effect.

Could you please apply the two attached patches to your 2.7 ? I've tested, they apply fine. I'm merging them into 2.8 but I'd prefer to get your feedback before backporting to 2.7, in case anything needs some adjustments. Thanks!

[0001-BUG-MINOR-sched-properly-report-long_rq-when-tasks-r.patch.txt](https://github.com/haproxy/haproxy/files/10759487/0001-BUG-MINOR-sched-properly-report-long_rq-when-tasks-r.patch.txt)
[0002-BUG-MEDIUM-sched-allow-a-bit-more-TASK_HEAVY-to-be-p.patch.txt](https://github.com/haproxy/haproxy/files/10759490/0002-BUG-MEDIUM-sched-allow-a-bit-more-TASK_HEAVY-to-be-p.patch.txt)

**Comment by kitkit201** [comment]
Hi @wtarreau I've been slammed with other work and didn't see your many replies. Many thanks in advance to you and your team for the thorough investigation! I will try these patches tomorrow US hours since it's already late here and i dont want to wake up the oncall if I break something. 

I did a check for the `ip_local_port_range` on the haproxy nodes and it appears to be near what you want already
```
~$ cat /proc/sys/net/ipv4/ip_local_port_range
1024	65023
```

Will be in touch with the patches! 

Cheers 🍻

**Comment by wtarreau** [comment]
OK so that's fine already and you can't do much more. It would be nice to have a look between haproxy and the server on the haproxy machine at any moment and more particularly during a next incident to see how many connections you find to the server, including those in TIME_WAIT if any. And if you find any in TIME_WAIT, it's important to estimate how many there are so that we understand how they are created, because you should have almost none.

**Comment by kitkit201** [comment]
Hi @wtarreau ,

As you saw in the update, I've went ahead and pushed out the patch version of HAProxxy 2.7.1 onto our corpus for it to marinate over the weekend. I tested it on a single host and it appears to be running normally and at least from initial reviews, we are using about the same amount of haproxy nodes in our corpus. 

I added some `sar -SOCK` and `ss -an` debug output to our script in case it comes up again. 

I'll update more as soon as I have something. 

Thanks <3

**Comment by wtarreau** [comment]
That's great, thanks for the feedback. By the way, please don't stay on 2.7.1 for too long, 2.7.3 is out and fixes a number of bugs and a vulnerability. I would prefer to know that you're running a fixed version!

**Comment by kitkit201** [comment]
Hahah yes!

I went ahead and upgraded all our hosts to 2.7.3 now with the patch. It's only 3 days old! :) nice! Have a good weekend! :)

**Comment by kitkit201** [comment]
Hi @wtarreau , 

I bring good news. We used the patch version of Haproxy over the past 10 days and did not see the same latency alerts we saw before the patch. 

We will likely slowly increase our CPU targeting as well to use less instances in due time. We will use this patch version until it gets pushed out to the next version of HAProxy 2.7 when it gets pushed out on your end. 

Thank you so much!

**Comment by wtarreau** [comment]
Cool, thanks for letting us know. We'll have a few other fixes that will be worth a 2.7.4, stay tuned ;-)

**Comment by wtarreau** [comment]
Now I've added in the stats the counter of number of times the global maxconn was reached (9615102b) . Thus the last point that was keeping this issue open was addressed, let's close it.

**Comment by mikkorantalainen** [comment]
> For some reason (attack, slow servers etc) at some points you're very likely exhausting the source port range and connect() starts to get slow, explaining these 6.7 ms to complete a loop.

Note that incoming connections are identified by tuple {client IP address, port number} so unless a single client IP creates more than ~20K connections HAProxy shouldn't be having problems finding a suitable port.

However, each `server` line in `backend` section causing *outgoing connection* to backend server eats one port from the server running HAProxy unless the same backend has multiple IP numbers. Limiting parallel connections from HAProxy to single backend IP to less than ~20K connections should avoid spending too much CPU to find suitable port number.

**Comment by git001** [comment]
> > For some reason (attack, slow servers etc) at some points you're very likely exhausting the source port range and connect() starts to get slow, explaining these 6.7 ms to complete a loop.
> 
> Note that incoming connections are identified by tuple {client IP address, port number} so unless a single client IP creates more than ~20K connections HAProxy shouldn't be having problems finding a suitable port.
> 
> However, each `server` line in `backend` section causing _outgoing connection_ to backend server eats one port from the server running HAProxy unless the same backend has multiple IP numbers. Limiting parallel connections from HAProxy to single backend IP to less than ~20K connections should avoid spending too much CPU to find suitable port number.

@mikkorantalainen please be so kind and open an new issue, as this is already closed and you setup could be different then in the origin post.
